mesonbuild/modules/python.py

```diff
@@ -0,0 +1,384 @@
+
+    @noPosargs
+    @noKwargs
+    def language_version(self, node, args, kwargs):
+        return ModuleReturnValue(self.version, [])
+
+    @noPosargs
+    @noKwargs
+    def found(self, node, args, kwargs):
+        return ModuleReturnValue(True, [])
+
+    @noKwargs
+    def get_path(self, node, args, kwargs):
+        if len(args) != 1:
+            raise InvalidArguments('get_path takes exactly one positional argument.')
+        path_name = args[0]
+        if not isinstance(path_name, str):
+            raise InvalidArguments('get_path argument must be a string.')
+
+        path = self.paths.get(path_name)
+
+        if path is None:
+            raise InvalidArguments('{} is not a valid path name'.format(path_name))
+
+        return ModuleReturnValue(path, [])
+
+    @noKwargs
+    def get_variable(self, node, args, kwargs):
+        if len(args) != 1:
+            raise InvalidArguments('get_variable takes exactly one positional argument.')
+        var_name = args[0]
+        if not isinstance(var_name, str):
+            raise InvalidArguments('get_variable argument must be a string.')
+
+        var = self.variables.get(var_name)
+
+        if var is None:
+            raise InvalidArguments('{} is not a valid path name'.format(var_name))
+
+        return ModuleReturnValue(var, [])
+
+    def method_call(self, method_name, args, kwargs):
+        try:
+            fn = getattr(self, method_name)
+        except AttributeError:
+            raise InvalidArguments('Python object does not have method %s.' % method_name)
+
+        if method_name in ['extension_module', 'dependency', 'install_sources']:
+            value = fn(self.interpreter, None, args, kwargs)
+            return self.interpreter.holderify(value)
+        else:
+            value = fn(None, args, kwargs)
+            return self.interpreter.module_method_callback(value)
+
+
+class PythonModule(ExtensionModule):
+    def __init__(self):
+        super().__init__()
+        self.snippets.add('find')
+
+    # https://www.python.org/dev/peps/pep-0397/
+    def _get_win_pythonpath(self, name_or_path):
+        if name_or_path not in ['python2', 'python3']:
+            return None
+        ver = {'python2': '-2', 'python3': '-3'}[name_or_path]
+        cmd = ['py', ver, '-c', "import sysconfig; print(sysconfig.get_config_var('BINDIR'))"]
+        _, stdout, _ = mesonlib.Popen_safe(cmd)
+        dir = stdout.strip()
+        if os.path.exists(dir):
+            return os.path.join(dir, 'python')
+        else:
+            return None
+
+    @permittedSnippetKwargs(['required'])
+    def find(self, interpreter, state, args, kwargs):
+        required = kwargs.get('required', True)
+        if not isinstance(required, bool):
+            raise InvalidArguments('"required" argument must be a boolean.')
+
+        name_or_path = args[0]
+        if not name_or_path:
+            python = ExternalProgram('python3', mesonlib.python_command, silent=True)
```

Comment:
Should this log something for the user to notice that this module might use a python shipped with meson instead of a python that is actually used on the system?
Can we detect if the python used was installed by the meson installer to have a more to the point log message?

Reply:
I don't know how one would go about detecting that to be honest :)

Reply:
I've added logging when going through that condition ("using meson's python {}")